### PR TITLE
[webview_flutter_android] Adds support for receiving Java callback `WebChromeClient.onShowFileChooser`

### PR DESCRIPTION
**Pull request by bparrishMines**
* Adds [WebChromeClient.onShowFileChooser](https://developer.android.com/reference/android/webkit/WebChromeClient#onShowFileChooser(android.webkit.WebView,%20android.webkit.ValueCallback%3Candroid.net.Uri[]%3E,%20android.webkit.WebChromeClient.FileChooserParams)) callback method.
* Adds [FileChooserParams](https://developer.android.com/reference/android/webkit/WebChromeClient.FileChooserParams) class since its passed to `onShowFileChooser`.
* Adds `AndroidWebViewController.setOnShowFileSelector`.
* Moves `WebChromeClient` from `AndroidNavigationDelegate` to `AndroidWebViewController` since the callback is set by the controller.

**Note**: This callback is only supported for versions 21+. 

Fixes https://github.com/flutter/flutter/issues/27924 and https://github.com/flutter/flutter/issues/97127

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> So, I added a helper method (`FileChooserParams.openFilePickerForResult()`) that handles starting an Activity and retrieving the result.
> 
> I was thinking that `openFilePickerForResult` could be used as the default functionality for Android. Therefore, a user wouldn't need the `file_picker` plugin to handle this.

I thought the decision from the previous discussion of option was that we would start with not providing a default, and suggest that people use `file_selector` as the implementation, so that we didn't have to maintain two copies of essentially the same code that we already have in `file_selector` (https://github.com/flutter/flutter/issues/27924#issuecomment-1040473753). Why the change in approach? Could we scope this PR down to just providing the callback part without a default implementation or helper, release that, and see if we actually need more that just an example of implementing it with `file_selector` (or maybe directly with `file_selector_android` to cut down on overhead to other platforms)?

**Comment by stuartmorgan** [comment]
I assume this PR will obsolete https://github.com/flutter/plugins/pull/5172?

**Comment by dupuchba** [comment]
Thank you both for your work 🙇

**Review by stuartmorgan — APPROVED** [review]
A couple of questions, but overall LGTM!

Once `file_selector_android` lands we should consider updating the example app for this package to show using it to handle this callback.

**Comment by rashidotm** [comment]
Thanks for the effort. I am looking forward to upgrade as soon as it is released. 🔥🔥

Approximately, what is the timeline on it?

**Comment by stuartmorgan** [comment]
As with all PRs, it will land when it's fully reviewed and approved.

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebChromeClientHostApiImpl.java [hunk 1]
```diff
@@ -45,6 +49,36 @@ public WebChromeClientImpl(@NonNull WebChromeClientFlutterApiImpl flutterApi) {
...
+        ValueCallback<Uri[]> filePathCallback,
+        FileChooserParams fileChooserParams) {
+      final boolean currentReturnValueForOnShowFileChooser = returnValueForOnShowFileChooser;
+      flutterApi.onShowFileChooser(
+          this,
+          webView,
+          fileChooserParams,
+          reply -> {
+            // The returned list of file paths can only be passed to `filePathCallback` if the
+            // `onShowFileChooser` method returned true.
```
[the remark this answers is not in the document]
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebChromeClientHostApiImpl.java:66]: Yea, I assumed there could be a situation where someone could want to know when a user clicks on the `<input name="userfile" type="file">` button. I don't think it hurts to keep it this way.

**Comment by kennetgo** [comment]
Can somebody please provide an example code to use setOnShowFileSelector, for an AndroidWebViewController. Thanks

**Comment by stuartmorgan** [comment]
@kennetgo https://github.com/flutter/flutter/issues/118836 is tracking that, as mentioned in the original issue.

**Comment by faccio02** [comment]
Still can't easily find an example for implementing image picker with `webview_flutter`

**Pull request merged**